pkg/api/types.go

```diff
@@ -2119,34 +2119,121 @@ const (
 	StrategicMergePatchType PatchType = "application/strategic-merge-patch+json"
 )
 
-// Type and constants for component health validation.
+// Type and constants for component type.
+type ComponentType string
+
+// ComponentType constants define some known types of components.
+// Each type may have multiple instances, as long as their names are unique.
+const (
+	ComponentAPIServer         ComponentType = "apiserver"
+	ComponentControllerManager ComponentType = "controller-manager"
+	ComponentScheduler         ComponentType = "scheduler" // TODO: move mesos-specific component type & validation to contrib
+)
+
+// ComponentSpec describes the attributes that a component is created with.
+type ComponentSpec struct {
+	// Type of the component
+	Type ComponentType `json:"type"`
+	// Address to reach the component (scheme, host, port, path). Address paths determined by component type.
+	Address string `json:"address,omitempty"`
```

Comment:
I would like to eventually break out the health-checking function into a separate API. 

I'd also like to leverage the existing Service and Endpoints APIs, rather than create yet another way to track application addresses, ports, names, etc.

I propose:
```
ServiceName string
ReadinessProbe *Probe // http probe host defaults to endpoint address; exec probes not supported
```

The ReadinessProbe would be optional, only needed to probe imported services.
A Service would need to be created for each component, but we plan to do that in the normal case, anyway. 
The combination of the Component and Endpoints would provide presence monitoring.

Self-registration would be a matter of updating the Endpoints resource corresponding to the component. Self-hosted components would not need to self-register.